wgpu-core/src/resource.rs

```diff
@@ -160,12 +160,47 @@ pub struct Buffer<B: hal::Backend> {
     pub(crate) device_id: Stored<DeviceId>,
     pub(crate) usage: wgt::BufferUsage,
     pub(crate) size: wgt::BufferAddress,
-    pub(crate) full_range: (),
+    // An allocated range in this allocator means that the range in question is NOT yet initialized.
+    pub(crate) uninitialized_ranges: range_alloc::RangeAllocator<wgt::BufferAddress>,
```

Comment:
I feel that `RangeAllocator` is a wrong choice here. The goal for the allocator is to automatically allocate and deallocate ranges (e.g. sub-allocation from a pool). But here instead we need quick intersection search, and we never use automatic allocation, since we know all the ranges...

One possible replacement would be `BTreeMap`, where the key is the start of a range. Since it's ordered by keys, it should be able to quickly find intersections. I'm just a little concerned about how we are going to be updating the range intersections there, merging, etc... 
Another possibility is to look for something like an "interval map" on crates.

Reply:
Agreed! I went with it because it was an idea that came up on the Matrix chat and I wasn't too keen on rewriting the merge logic. With RangeAllocator we also necessarily have this weird list of uninitialized ranges instead of the other way round because only free takes ranges on free.

Since you, as implied, don't mind adding another crate for this (if it doesn't have a log of dependencies I reckon :)) I'll have a look around for that. Otherwise, I may bite the bullet after all and write a dedicated thing either using BTreeMap/similar or from scratch which then can be used both here and on the command buffers.

Reply:
Sorry about misleading you into `RangeAllocator`! This isn't one of the tasks that's obvious, it requires a bit of R&D, and I believe you are fully capable of doing it (assuming you have the time/fuel to invest).